### PR TITLE
Work branch

### PR DESCRIPTION
**Pull request by loresfca**

### CODE REVIEW

File: Week 4/Generate Parenthesis/Solution.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import java.io.*;
```
Comment (charientism) [inline Week 4/Generate Parenthesis/Solution.java:1]: Wow! You went above and beyond this week by tackling the bonus problems in addition to your homework! Awesome job :)

File: Week 4/Word Break/TEBOW.txt [hunk 2]
```diff
@@ -0,0 +1,41 @@
+Word Break:
```
Comment (charientism) [inline Week 4/Word Break/TEBOW.txt:1]: Thanks for uploading your TEBOW IT! You did an especially good job of picking small test cases that are easy to walkthrough and taking care to consider edge cases.

File: Week 4/Word Break/TEBOW.txt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+Complexity 2^n
+
+Optimization:
+Store the strings previously seen and the result they give. 
+Complexity n^2
+
+WalkThrough:
+1.- Input validation
+2.- Check recursively the prefix and the suffix of the string and store seen strings
+4.- return result
```
Comment (charientism) [inline Week 4/Word Break/TEBOW.txt:41]: Implement: Don't forget to code up your implementation via pencil and paper or on a whiteboard if you can. It's important to simulate the way your real interview would go as closely as possible while you're practicing and time yourself to 45 minutes. (After that, you can code everything up/test on the computer.)

File: Week 4/Word Break/Solution.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    	if(s.length()==0||dict.length==0) return false;
+    	//To store previously checked suffixes.
+    	HashMap<String,Boolean> suffix = new HashMap<String,Boolean>();
+    	//Start of recursion.
+    	return validDict(s,dict,suffix);
+    }
+    public boolean validDict(String s,String[] dict,HashMap<String,Boolean> suffix){
+    	//for manages the size of the prefix
+    	for(int i = 1; i<=s.length();i++){
+    		//prefix
```
Comment (charientism) [inline Week 4/Word Break/Solution.java:20]: Thanks for commenting :)

File: Week 4/Word Break/Solution.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    				}
+    			}else{
+    				//if the current string doesn't have a suffix and is valid true is returned.
+    				return true;
+    			}
+    		}
+    	}
+    	return false;
+    }
+    public boolean contains(String[] dict,String curr){
```
Comment (charientism) [inline Week 4/Word Break/Solution.java:46]: Minor optimization - convert the search string + dictionary strings to lowercase in the setup function. This reduces the number of operations you're doing on any given search of the dictionary (even though it doesn't change your Big O).

File: Week 4/Word Break/Solution.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
+import java.io.*;
+import java.util.*;
+import org.junit.Test;
+import static org.junit.Assert.assertEquals;
+public class Solution {
+    public boolean validDict(String s,String[] dict) {
+    	if(s==null||dict==null) throw new IllegalArgumentException();
```
Comment (charientism) [inline Week 4/Word Break/Solution.java:7]: Good job setting up the helper method and using it to set up your helper data structure and do input validation - that's exactly what you'll want to do in your interviews.

**Review by charientism — APPROVED** [review]

File: Week 4/Word Break/TEBOW.txt (hunk 3 above)
Comment (loresfca) [reply Week 4/Word Break/TEBOW.txt:41]: Thanks, will do!

File: Week 4/Word Break/Solution.java (hunk 5 above)
Comment (loresfca) [reply Week 4/Word Break/Solution.java:46]: Thanks, I'll implement that
